sdk/python/kfp/containers/_build_image_api.py

```diff
@@ -0,0 +1,148 @@
+
+    # Creating requirements.txt from the list of installed packages if requirements.txt does not already exist.
+    if not requirements_file_exists:
+        logging.error('''requirements.txt file was not found''')
+        generate_requirements_file(requirements_path)
+        requirements_file_exists = True
+
+    base_image = default_base_image
+    if callable(base_image):
+        base_image = base_image()
+
+    dockerfile_lines = []
+    dockerfile_lines.append('FROM {}'.format(base_image))
+    dockerfile_lines.append('COPY . {}'.format(_container_work_dir))
+    dockerfile_lines.append('WORKDIR {}'.format(_container_work_dir))
+    if requirements_file_exists:
+        dockerfile_lines.append('RUN python3 -m pip install -r {}'.format(requirements_rel_path))
+
+    return '\n'.join(dockerfile_lines)
+
+
+def build_image_from_env(image_name : str = None, source_dir : str = None, file_filter_re : str = '.*\.py',  timeout : int = 1000, builder : ContainerBuilder = None):
+    '''build_image_from_env builds and pushes a new container image that reconstructs the current python environment.
+    Args:
+        image_name: Optional. The image repo name where the new container image will be pushed. The name will be generated if not not set.
+        source_dir: Optional. The directory that will be used as the environment source. The requirements.txt and python files inside the directory will be captured.
+        timeout: Optional. The image building timeout in seconds.
+        builder: Optional. An instance of ContainerBuilder or compatible class that will be used to build the image.
+    '''
+    current_dir = source_dir or os.getcwd()
+    with tempfile.TemporaryDirectory() as context_dir:
+        logging.info('Creating the build context directory: {}'.format(context_dir))
+
+        # Copying all *.py and requirements.txt files
+        for dirpath, dirnames, filenames in os.walk(current_dir):
+            dst_dirpath = os.path.join(context_dir, os.path.relpath(dirpath, current_dir))
+            os.makedirs(dst_dirpath, exist_ok=True)
+            for file_name in filenames:
+                if re.match(file_filter_re, file_name) or file_name == 'requirements.txt':
+                    src_path = os.path.join(dirpath, file_name)
+                    dst_path = os.path.join(dst_dirpath, file_name)
+                    shutil.copy(src_path, dst_path)
+            
+
+        src_dockerfile_path = os.path.join(current_dir, 'Dockerfile')
+        dst_dockerfile_path = os.path.join(context_dir, 'Dockerfile')
+        if os.path.exists(src_dockerfile_path):
+            shutil.copy(src_dockerfile_path, dst_dockerfile_path)
+        else:
+            dockerfile_text = generate_dockerfile_text(context_dir, dst_dockerfile_path)
+            with open(dst_dockerfile_path, 'w') as f:
+                f.write(dockerfile_text)
+
+        if builder is None:
+            builder = get_default_image_builder()
+        return builder.build(
```

Comment:
Maybe not in this PR. We will need add a simple local cache to avoid rebuilding images every-time if there is no code change.

Reply:
I agree with the importance of caching, but think it should be handled on the lower builder level.